jgrapht-ext/src/main/java/org/jgrapht/ext/GmlImporter.java

```diff
@@ -225,36 +229,62 @@ public void syntaxError(
         private Integer sourceId;
         private Integer targetId;
         private Double weight;
-
-        private Set<Integer> nodes;
-        private int singletons;
+        private String label;
+
+        /**
+         * Maps IDs, which are always integers, to the corresponding node labels
+         * for nodes that are connected in the graph.
+         */
+        private HashMap<Integer, String> nodeLabels;
+        /** Labels for stand-alone nodes that are not connected.  Note that these
+         * can be NULL.
+         */
+        private List<String> singletonLabels;
         private List<PartialEdge> edges;
 
         public void updateGraph(Graph<V, E> graph)
             throws ImportException
         {
             if (foundGraph) {
-                // add nodes
+                // Add connected nodes
                 int maxV = 1;
                 Map<Integer, V> map = new HashMap<Integer, V>();
-                for (Integer id : nodes) {
+                for (Integer id : nodeLabels.keySet()) {
+                	String label = nodeLabels.get(id);
+                	// Substitute the ID for the label if no label provided
+                	if (label == null || (label != null && label.equals(""))) {
+                		label = id + "";
+                	}
                     maxV = Math.max(maxV, id);
-                    V vertex =
-                        vertexProvider.buildVertex(id.toString(), new HashMap<String, String>());
-                    map.put(id, vertex);
-                    graph.addVertex(vertex);
+                    V vertex = vertexProvider.buildVertex(label, new HashMap<String, String>());
+                    if (vertex != null) {
+                    	map.put(id, vertex);
+                    	graph.addVertex(vertex);
+                    }
                 }
 
-                // add singleton nodes
-                for (int i = 0; i < singletons; i++) {
-                    String label = String.valueOf(maxV + 1 + i);
-                    graph.addVertex(
-                        vertexProvider.buildVertex(label, new HashMap<String, String>()));
+                // Add singleton nodes--those with no IDs and that are not connected.
+                for (int i = 0; i < singletonLabels.size(); i++) {
+                	String label = singletonLabels.get(i);
+                	// For nodes that for whatever reason do not have IDs set in the file,
+                	// assign the next higher ID as the label.
+                	if (label == null || (label != null && label.equals(""))) {
+                		label = String.valueOf(maxV + 1 + i);
+                	}
+                	V vertex = vertexProvider.buildVertex(label, new HashMap<String, String>());
```

Comment:
Same here, first parameter the id and only if there is a label, the label inside the `HashMap`.